File: db/dataParsingAlpha/parsing/csvRun.go
package main

import (
	"database/sql"
	"encoding/csv"
	"fmt"
	"github.com/lib/pq"
	"io"
	"log"
	"os"
	"regexp"
	"strings"
	"time"
)

var providerColumns = [...]string{
	"entity_type_code",
	"replacement_npi",
	"last_name_legal_name",
	"first_name",
	"middle_name",
	"name_prefix",
	"name_suffix",
	"credential",
	"other_last_name",
	"other_first_name",
	"other_middle_name",
	"other_name_prefix",
	"other_name_suffix",
	"other_credential",
	"other_last_name_type_code",
	"enumeration_date",
	"last_update_date",
	"npi_deactivation_reason_code",
	"npi_deactivation_date",
	"npi_reactivation_date",
	"gender_code",
	"is_sole_proprietor",
	"npi"}
var providerHeaders = [...]string{"entity_type_code", "replacement_npi",
	"provider_last_name", "provider_first_name",
	"provider_middle_name", "provider_name_prefix",
	"provider_name_suffix", "provider_credential",
	"provider_other_last_name", "provider_other_first_name",
	"provider_other_middle_name", "provider_other_name_prefix",
	"provider_other_name_suffix", "provider_other_credential",
	"provider_other_last_name_type_code", "provider_enumeration_date",
	"last_update_date", "npi_deactivation_reason_code",
	"npi_deactivation_date", "npi_reactivation_date",
	"provider_gender_code",
	"is_sole_proprietor"}

var organizationColumns = [...]string{
	"entity_type_code",
	"replacement_npi",
	"ein",
	"organization_name_legal_business_name",
	"other_organization_name",
	"other_organization_name_type_code",
	"enumeration_date",
	"last_update_date",
	"npi_deactivation_code",
	"npi_deactivation_date",
	"npi_reactivation_date",
	"authorized_official_last_name",
	"authorized_official_first_name",
	"authorized_official_middle_name",
	"authorized_official_name_prefix",
	"authorized_official_name_suffix",
	"authorized_official_credential",
	"authorized_official_titleor_position",
	"authorized_official_telephone_number",
	"is_organization_subpart",
	"parent_organization_lbn",
	"parent_organization_tin",
	"npi"}
var organizationHeaders = [...]string{
	"entity_type_code",
	"replacement_npi",
	"employer_identification_number",
	"provider_organization_name",
	"provider_other_organization_name",
	"provider_other_organization_name_type_code",
	"provider_enumeration_date",
	"last_update_date",
	"npi_deactivation_reason_code",
	"npi_deactivation_date",
	"npi_reactivation_date",
	"authorized_official_last_name",
	"authorized_official_first_name",
	"authorized_official_middle_name",
	"authorized_official_name_prefix",
	"authorized_official_name_suffix",
	"authorized_official_credential",
	"authorized_official_titleor_position",
	"authorized_official_telephone_number",
	"is_organization_subpart",
	"parent_organization_lbn",
	"parent_organization_tin"}

var addressColumns = [...]string{
	"type", // mailing_address or practice_location_address
	"first_line",
	"second_line",
	"city",
	"state",
	"postal_code",
	"country_code",
	"telephone_number",
	"fax_number",
	"entity_type",
	"entity_id"}
var mailingHeaders = [...]string{"provider_first_line_business_mailing_address", "provider_second_line_business_mailing_address",
	"provider_business_mailing_address_city_name", "provider_business_mailing_address_state_name",
	"provider_business_mailing_address_postal_code", "provider_business_mailing_address_country_code",
	"provider_business_mailing_address_telephone_number", "provider_business_mailing_address_fax_number"}
var practiceLocationHeaders = [...]string{"provider_first_line_business_practice_location_address", "provider_second_line_business_practice_location_address",
	"provider_business_practice_location_address_city_name", "provider_business_practice_location_address_state_name",
	"provider_business_practice_location_address_postal_code", "provider_business_practice_location_address_country_code",
	"provider_business_practice_location_address_telephone_number", "provider_business_practice_location_address_fax_number"}

var otherProviderColumns = [...]string{
	"identifier",
	"identifier_type_code",
	"identifier_state",
	"identifier_issuer",
	"entity_type",
	"entity_id"}
var otherProviderHeaders = [...]string{"other_provider_identifier_%d", "other_provider_identifier_type_code_%d", "other_provider_identifier_state_%d", "other_provider_identifier_issuer_%d"}

var taxonomyLicenseColumns = [...]string{
	"taxonomy_code",
	"license_number",
	"license_number_state_code",
	"primary_taxonomy_switch",
	"entity_type",
	"entity_id"}
var taxonomyLicenseHeaders = [...]string{"healthcare_provider_taxonomy_code_%d", "provider_license_number_%d", "provider_license_number_state_code_%d", "healthcare_provider_primary_taxonomy_switch_%d"}

var taxonomyGroupColumns = [...]string{
	"taxonomy_group",
	"entity_type",
	"entity_id"}
var taxonomyGroupHeaders = [...]string{"healthcare_provider_taxonomy_group_%d"}

func underscore(string string) string {
	parensRegexp := regexp.MustCompile(" \\(.*\\)")
	string = parensRegexp.ReplaceAllString(string, "")
	string = strings.ToLower(string)
	return strings.Replace(string, " ", "_", -1)
}

// notAllNull feels hacky, but should allow us to remove completely null values from the inserts
func insertData(headers []string, recordMap map[string]string, stmt *sql.Stmt, repeat int) {
	for i := 1; i <= repeat; i++ {
		var values []interface{}
		for _, header := range headers {
			if strings.Contains(header, "%d") {
				header = fmt.Sprintf(header, i)
			}
			value := recordMap[header]
			if len(value) > 0 {
				values = append(values, value)
			} else {
				values = append(values, nil)
			}
		}
		values = append(values, recordMap["npi"])
		_, err := stmt.Exec(values...)
		if err != nil {
			fmt.Println(values)
			log.Fatal(err)
		}
	}
	//fmt.Println(values)
}

func insert(recordMap map[string]string, stmts map[string]*sql.Stmt) {
	if recordMap["entity_type_code"] == "1" {
		insertData(providerHeaders[:], recordMap, stmts["providers"], 1)
		recordMap["entity_type"] = "Provider"
	} else if recordMap["entity_type_code"] == "2" {
		insertData(organizationHeaders[:], recordMap, stmts["organizations"], 1)
		recordMap["entity_type"] = "Organization"
	} else if recordMap["entity_type_code"] == "" {
	} else {
		log.Fatal("Unknown entity type code")
	}

	recordMap["address_type"] = "MailingAddress"
	mailingHeaders := append([]string{"address_type"}, mailingHeaders[:]...) //prepend address_type
	mailingHeaders = append(mailingHeaders, "entity_type")                   // append entity_type
	insertData(mailingHeaders[:], recordMap, stmts["addresses"], 1)

	recordMap["address_type"] = "PracticeLocationAddress"
	practiceLocationHeaders := append([]string{"address_type"}, practiceLocationHeaders[:]...)
	practiceLocationHeaders = append(practiceLocationHeaders, "entity_type")
	insertData(practiceLocationHeaders[:], recordMap, stmts["addresses"], 1)

	otherProviderHeaders := append(otherProviderHeaders[:], "entity_type")
	insertData(otherProviderHeaders[:], recordMap, stmts["other_provider_identifiers"], 50)

	taxonomyGroupHeaders := append(taxonomyGroupHeaders[:], "entity_type")
	insertData(taxonomyGroupHeaders[:], recordMap, stmts["taxonomy_groups"], 15)

	taxonomyLicenseHeaders := append(taxonomyLicenseHeaders[:], "entity_type")
	insertData(taxonomyLicenseHeaders[:], recordMap, stmts["taxonomy_licenses"], 15)

}

func prepareCopyIn(table string, columns []string, db *sql.DB) (*sql.Stmt, *sql.Tx) {
	txn, err := db.Begin()
	if err != nil {
		log.Fatal(err)
	}
	copyIn := pq.CopyIn(table, columns...)
	output, err := txn.Prepare(copyIn)
	if err != nil {
		fmt.Println(table)
		log.Fatal(err)
	}
	return output, txn
}

func main() {

	startingTime := time.Now()
	fmt.Println("Started parsing", startingTime)

	csvfile, err := os.Open(os.Args[2])
	if err != nil {
		log.Fatal(err)
	}
	defer csvfile.Close()

	csvReader := csv.NewReader(csvfile)
	fmt.Println("reading csv parsing", time.Now())

	headers, err := csvReader.Read()
	for idx, header := range headers {
		headers[idx] = underscore(header)
	}
	if err != nil {
		log.Fatal(err)
	}

	db, err := sql.Open("postgres", " dbname="+os.Args[1]+" sslmode=disable")
	txns := make(map[string]*sql.Tx)
	copyIn := make(map[string]*sql.Stmt)
	copyIn["providers"], txns["providers"] = prepareCopyIn("providers", providerColumns[:], db)
	copyIn["organizations"], txns["organizations"] = prepareCopyIn("organizations", organizationColumns[:], db)
	copyIn["addresses"], txns["addresses"] = prepareCopyIn("addresses", addressColumns[:], db)
	copyIn["other_provider_identifiers"], txns["other_provider_identifiers"] = prepareCopyIn("other_provider_identifiers", otherProviderColumns[:], db)
	copyIn["taxonomy_licenses"], txns["taxonomy_licenses"] = prepareCopyIn("taxonomy_licenses", taxonomyLicenseColumns[:], db)
	copyIn["taxonomy_groups"], txns["taxonomy_groups"] = prepareCopyIn("taxonomy_groups", taxonomyGroupColumns[:], db)

	for {

		record, err := csvReader.Read()

		if err != nil {
			if err == io.EOF {
				break
			}
			log.Fatalln(err)
		}

		recordMap := make(map[string]string)
		for idx, value := range record {
			if len(value) > 0 {
				recordMap[headers[idx]] = value
			}
		}

		insert(recordMap, copyIn)
	}

	for _, stmt := range copyIn {
		err = stmt.Close()
		if err != nil {
			log.Fatal(err)
		}
	}
	for _, txn := range txns {
		err = txn.Commit()
		if err != nil {
			log.Fatal(err)
		}
	}

	fmt.Println("Finished parsing", time.Now())
}

File: db/dataParsingAlpha/parsing/csvRun.rb
# Takes two arguments, first is the database name where the data should be loaded, and the second is the file path to the NPPES data

require 'time'
require 'pg'
require 'csv'

conn = PG.connect( dbname: ARGV[0] )
createProvider = "CREATE TABLE IF NOT EXISTS provider (
  npi bigint PRIMARY KEY,
  entity_type_code int,
  replacement_npi bigint,
  ein varchar,
  provider_organization_name_legal_business_name varchar,
  provider_last_name_legal_name varchar,
  provider_first_name varchar,
  provider_middle_name varchar,
  provider_name_prefix varchar,
  provider_name_suffix varchar,
  provider_credential varchar,
  provider_other_organization_name varchar,
  provider_other_organization_name_type_code varchar,
  provider_other_last_name varchar,
  provider_other_first_name varchar,
  provider_other_middle_name varchar,
  provider_other_name_prefix varchar,
  provider_other_name_suffix varchar,
  provider_other_credential varchar,
  provider_other_last_name_type_code bigint,
  provider_enumeration_date date,
  last_update_date date,
  npi_deactivation_reason_code varchar,
  npi_deactivation_date date,
  npi_reactivation_date date,
  provider_gender_code varchar,
  authorized_official_last_name varchar,
  authorized_official_first_name varchar,
  authorized_official_middle_name varchar,
  authorized_official_titleor_position varchar,
  authorized_official_telephone_number varchar,
  is_sole_proprietor varchar,
  is_organization_subpart varchar,
  parent_organization_lbn varchar,
  parent_organization_tin varchar,
  authorized_official_name_prefix varchar,
  authorized_official_name_suffix varchar,
  authorized_official_credential varchar
);"
insertProvider = "INSERT INTO provider VALUES ("

createAddress = "CREATE TABLE IF NOT EXISTS address (
  type VARCHAR,
  first_line VARCHAR,
  second_line VARCHAR,
  city_name VARCHAR,
  state_name VARCHAR,
  postal_code VARCHAR,
  country_code VARCHAR,
  telephone_number VARCHAR,
  fax_number VARCHAR,
  npi_fk bigint references provider(npi)
);"
insertAddress = "INSERT INTO address VALUES ("

createHealthcareProviderTaxonomyGroup = "CREATE TABLE IF NOT EXISTS healthcare_provider_taxonomy_group (
  taxonomy_group varchar,
  npi_fk bigint references provider(npi)
);"
insertHealthcareProviderTaxonomyGroup = "INSERT INTO healthcare_provider_taxonomy_group VALUES ("

createOtherProvider = "CREATE TABLE IF NOT EXISTS other_provider (
  other_provider_identifier varchar,
  other_provider_identifier_type_code varchar,
  other_provider_identifier_state varchar,
  other_provider_identifier_issuer varchar,
  npi_fk bigint references provider(npi)
);"
insertOtherProvider = "INSERT INTO other_provider VALUES ("

createTaxonomyTable = "CREATE TABLE IF NOT EXISTS taxonomy_license (
  healthcare_provider_taxonomy_code varchar,
  provider_license_number varchar,
  provider_license_number_state_code varchar,
  healthcare_provider_primary_taxonomy_switch varchar,
  npi_fk bigint references provider(npi)
);"
insertTaxonomyTable = "INSERT INTO taxonomy_license VALUES ("

def nullEmptyCheck(val)
  val.gsub!("'", "''")
  output = (!val || val.empty?) ? "DEFAULT" : "'"+val+"'"
  return output
end

def createInsert(csvIndicies, row, reportNull = false)
  output = ""
  nonNullCount = 0
  csvIndicies.each_with_index do |indicie, index|
    ending = index == csvIndicies.length-1 ? '' : ','
    deafultCheck = nullEmptyCheck(row[indicie])
    nonNullCount = nonNullCount + 1 if deafultCheck == "DEFAULT"
    output = output + deafultCheck + ending
  end
  if (nonNullCount == csvIndicies.length-1 && reportNull)
    output = nil
  end
  return output
end
puts Time.now
i = 0

conn.exec("DROP TABLE IF EXISTS address")
conn.exec("DROP TABLE IF EXISTS healthcare_provider_taxonomy_group")
conn.exec("DROP TABLE IF EXISTS other_provider")
conn.exec("DROP TABLE IF EXISTS taxonomy_license")
conn.exec("DROP TABLE IF EXISTS provider")

conn.exec(createProvider)
conn.exec(createAddress)
conn.exec(createHealthcareProviderTaxonomyGroup)
conn.exec(createOtherProvider)
conn.exec(createTaxonomyTable)
output = ""
CSV.foreach(ARGV[1], :headers => true) do |row| 
  i = i + 1
  output = output + insertProvider+
  (
    createInsert(["NPI","Entity Type Code",
      "Replacement NPI",
      "Employer Identification Number (EIN)",
      "Provider Organization Name (Legal Business Name)",
      "Provider Last Name (Legal Name)",
      "Provider First Name",
      "Provider Middle Name",
      "Provider Name Prefix Text",
      "Provider Name Suffix Text",
      "Provider Credential Text",
      "Provider Other Organization Name",
      "Provider Other Organization Name Type Code",
      "Provider Other Last Name",
      "Provider Other First Name",
      "Provider Other Middle Name",
      "Provider Other Name Prefix Text",
      "Provider Other Name Suffix Text",
      "Provider Other Credential Text",
      "Provider Other Last Name Type Code",
      "Provider Enumeration Date",
      "Last Update Date",
      "NPI Deactivation Reason Code",
      "NPI Deactivation Date",
      "NPI Reactivation Date",
      "Provider Gender Code",
      "Authorized Official Last Name",
      "Authorized Official First Name",
      "Authorized Official Middle Name",
      "Authorized Official Title or Position",
      "Authorized Official Telephone Number",
      "Is Sole Proprietor",
      "Is Organization Subpart",
      "Parent Organization LBN",
      "Parent Organization TIN",
      "Authorized Official Name Prefix Text",
      "Authorized Official Name Suffix Text",
      "Authorized Official Credential Text"
    ], row) + ');'
  )+
  insertAddress+ "'mailing'," + 
  (
    createInsert(["Provider First Line Business Mailing Address","Provider Second Line Business Mailing Address",
    "Provider Business Mailing Address City Name","Provider Business Mailing Address State Name",
    "Provider Business Mailing Address Postal Code","Provider Business Mailing Address Country Code (If outside U.S.)",
    "Provider Business Mailing Address Telephone Number","Provider Business Mailing Address Fax Number","NPI"], row) + ');'
  )+
  insertAddress+ "'practice_location'," +
  (
    createInsert(["Provider First Line Business Practice Location Address","Provider Second Line Business Practice Location Address",
    "Provider Business Practice Location Address City Name","Provider Business Practice Location Address State Name",
    "Provider Business Practice Location Address Postal Code","Provider Business Practice Location Address Country Code (If outside U.S.)",
    "Provider Business Practice Location Address Telephone Number","Provider Business Practice Location Address Fax Number","NPI"], row) + ');'
  )
  for j in 1..15
    toInsert = createInsert(["Healthcare Provider Taxonomy Group_#{j}","NPI"], row, true) 
    if toInsert == nil
      next
    end
    output = output + insertHealthcareProviderTaxonomyGroup + toInsert + ');'
  end
  for j in 1..50
    toInsert = createInsert(["Other Provider Identifier_#{j}", "Other Provider Identifier Type Code_#{j}", 
      "Other Provider Identifier State_#{j}", "Other Provider Identifier Issuer_#{j}","NPI"], row, true) 
    if toInsert == nil
      next
    end
    output = output + insertOtherProvider + toInsert + ');'
  end
  for j in 1..15
    toInsert = createInsert(["Healthcare Provider Taxonomy Code_#{j}", "Provider License Number_#{j}", 
      "Provider License Number State Code_#{j}", "Healthcare Provider Primary Taxonomy Switch_#{j}","NPI"], row, true) 
    if toInsert == nil
      next
    end
    output = output + insertOtherProvider + toInsert + ');'
  end
  if i == 40

    conn.exec(output)
    output = ""
    i = 0
  end

end
conn.exec(output)
puts Time.now
puts i
